### PR TITLE
feat: android 14 permission support

### PR DESCRIPTION
**Pull request by matinzd**
BREAKING CHANGE: You will need to make some changes to your `MainActivity.kt` file

Blocked by: https://github.com/facebook/react-native/issues/42403

### CODE REVIEW

File: example/android/app/src/main/java/com/healthconnectexample/PermissionRationaleActivity.kt [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import android.webkit.WebResourceRequest
+import android.webkit.WebView
+import android.webkit.WebViewClient
+import androidx.appcompat.app.AppCompatActivity
+
+class PermissionsRationaleActivity: AppCompatActivity() {
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+
+    val webView = WebView(this)
```
Comment (TheRogue76) [inline example/android/app/src/main/java/com/healthconnectexample/PermissionRationaleActivity.kt:13]: Why are you adding a Webview in Example app that shows the getting started guide for health-connect?

File: example/android/app/src/main/java/com/healthconnectexample/MainActivity.kt [hunk 2]
```diff
@@ -1,22 +1,29 @@
...
   /**
    * Returns the name of the main component registered from JavaScript. This is used to schedule
    * rendering of the component.
    */
   override fun getMainComponentName(): String = "HealthConnectExample"
 
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    // In order to handle contract results, we need to set the permission delegate.
+    HealthConnectPermissionDelegate.setPermissionDelegate(this)
```
Comment (TheRogue76) [inline example/android/app/src/main/java/com/healthconnectexample/MainActivity.kt:20]: Doc this. And maybe add a little Migration guide or something in the ReadMe.

Comment (matinzd) [reply example/android/app/src/main/java/com/healthconnectexample/MainActivity.kt:20]: Yup! Will do!

File: example/android/app/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -22,9 +22,29 @@
...
+        android:name=".PermissionsRationaleActivity"
+        android:exported="true">
         <intent-filter>
           <action android:name="androidx.health.ACTION_SHOW_PERMISSIONS_RATIONALE" />
         </intent-filter>
       </activity>
+
+      <!-- For versions starting Android 14, create an activity alias to show the rationale
+           of Health Connect permissions once users click the privacy policy link. -->
+      <activity-alias
```
Comment (TheRogue76) [inline example/android/app/src/main/AndroidManifest.xml:39]: Where do i need to sign a petition to fire devs in Google who change APIs for Shits and giggles?

File: example/android/app/src/main/java/com/healthconnectexample/PermissionRationaleActivity.kt (hunk 1 above)
Comment (matinzd) [reply example/android/app/src/main/java/com/healthconnectexample/PermissionRationaleActivity.kt:13]: Yeah. It's a sample activity for when user clicks on a privacy policy link inside health connect.

File: example/android/app/src/main/AndroidManifest.xml (hunk 3 above)
Comment (matinzd) [reply example/android/app/src/main/AndroidManifest.xml:39]: Make something alpha for 3 years. Take no responsibility and do breaking changes all the time.

File: android/src/main/java/dev/matinzd/healthconnect/permissions/HealthConnectPermissionDelegate.kt [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+import kotlinx.coroutines.launch
+
+object HealthConnectPermissionDelegate {
+  private lateinit var requestPermission: ActivityResultLauncher<Set<String>>
+  private val channel = Channel<Set<String>>()
+  private val coroutineScope = CoroutineScope(Dispatchers.IO)
+
+  fun setPermissionDelegate(
+    activity: ReactActivity,
+    providerPackageName: String? = "com.google.android.apps.healthdata"
```
Comment (TheRogue76) [inline android/src/main/java/dev/matinzd/healthconnect/permissions/HealthConnectPermissionDelegate.kt:19]: Why `String?` ? You need the providerPackageName, so just mark it as a `String` and give it this default. You can then remove the force unwrap on line 21:
```kt
PermissionController.createRequestPermissionResultContract(providerPackageName)
```

**Review by TheRogue76 — COMMENTED** [review]
Left some comments. This is why my side projects never have Android. If the API is gonna be shitty, make it non-depracatable like web, and if it is depracatable make it not shitty like iOS. Android does neither

File: android/src/main/java/dev/matinzd/healthconnect/permissions/HealthConnectPermissionDelegate.kt (hunk 4 above)
Comment (matinzd) [reply android/src/main/java/dev/matinzd/healthconnect/permissions/HealthConnectPermissionDelegate.kt]: That's write. My bad. Fixed under [145fa8a](https://github.com/matinzd/react-native-health-connect/pull/60/commits/145fa8a75b6c25315191b4f6c0d989672cc6d05e).

**Comment by matinzd** [comment]
Waiting for https://github.com/facebook/react-native/pull/42478 to be merged.

**Comment by TheRogue76** [comment]
Unblocked. Go for it

**Review by TheRogue76 — APPROVED** [review]

**Comment by matinzd** [comment]
We need to wait for release crew to start back-porting this fix to older versions of RN:

RN 70: https://github.com/reactwg/react-native-releases/discussions/92#discussioncomment-8208813 
RN 71: https://github.com/reactwg/react-native-releases/discussions/98#discussioncomment-8208768
RN 72: https://github.com/reactwg/react-native-releases/discussions/99#discussioncomment-8208809
RN 73: https://github.com/reactwg/react-native-releases/discussions/100#discussioncomment-8208796

**Comment by mgarcialeniolabs** [comment]
Any updates on this? Can see that on RN the changes has been merged.

**Comment by matinzd** [comment]
> Any updates on this? Can see that on RN the changes has been merged.

I am still waiting for the backport fix for 72.

**Review by TheRogue76 — APPROVED** [review]

**Comment by icernigoj** [comment]
Hey there 

I'm really looking forward to this update. Is there any way I can help move things along?

**Comment by KesoGizmoYoshi** [comment]
> Hey there
> 
> I'm really looking forward to this update. Is there any way I can help move things along?

I think you just have to wait for React Native 0.72.11 to be released.

**Comment by abdullatifHH** [comment]
+1
@matinzd is there any update on this.

**Comment by matinzd** [comment]
Waiting for RN 72 patch. Until then, no updates.

**Comment by mmeinzer** [comment]
@matinzd We upgraded to the latest version of RN 73 and this package with the Android 14 fix but get the error that seems related to the update mentioned in the `Readme` to `MainActivity.kt`. We don't currently maintain any custom native code in our project (we use Expo) - is that something we will need to do in order to get v2 of `react-native-health-connect` working?

**Comment by matinzd** [comment]
> @matinzd We upgraded to the latest version of RN 73 and this package with the Android 14 fix but get the error that seems related to the update mentioned in the `Readme` to `MainActivity.kt`. We don't currently maintain any custom native code in our project (we use Expo) - is that something we will need to do in order to get v2 of `react-native-health-connect` working? 

I need to publish a version that uses expo modules api😅

**Pull request merged**